airbyte-integrations/connectors/source-microsoft-onedrive/source_microsoft_onedrive/stream_reader.py

```diff
@@ -95,55 +113,133 @@ def config(self, value: SourceMicrosoftOneDriveSpec):
         assert isinstance(value, SourceMicrosoftOneDriveSpec)
         self._config = value
 
+    @property
+    @lru_cache(maxsize=None)
+    def drives(self):
+        """
+        Retrieves and caches OneDrive drives, including the user's drive based on authentication type.
+        """
+        drives = self.one_drive_client.drives.get().execute_query()
+
+        if self.config.credentials.auth_type == "Client":
+            my_drive = self.one_drive_client.me.drive.get().execute_query()
+        else:
+            my_drive = (
+                self.one_drive_client.users.get_by_principal_name(self.config.credentials.user_principal_name).drive.get().execute_query()
+            )
+
+        drives.add_child(my_drive)
+
+        # filter only onedrive drives
+        drives = list(filter(lambda drive: drive.drive_type in ["personal", "business"], drives))
+
+        return drives
+
+    def _get_shared_drive_object(self, drive_id: str, object_id: str, path: str) -> List[Tuple[str, str, datetime]]:
+        """
+        Retrieves a list of all nested files under the specified object.
+        Args:
+            drive_id: The ID of the drive containing the object.
+            object_id: The ID of the object to start the search from.
+        Returns:
+            A list of tuples containing file information (name, download URL, and last modified datetime).
+        Raises:
+            RuntimeError: If an error occurs during the request.
+        """
+
+        access_token = self.get_access_token()
+        headers = {"Authorization": f"Bearer {access_token}"}
+        base_url = f"https://graph.microsoft.com/v1.0/drives/{drive_id}"
+
+        def get_files(url: str, path: str) -> List[Tuple[str, str, datetime]]:
+            response = requests.get(url, headers=headers)
+            if response.status_code != 200:
+                error_info = response.json().get("error", {}).get("message", "No additional error information provided.")
+                raise RuntimeError(f"Failed to retrieve files from URL '{url}'. HTTP status: {response.status_code}. Error: {error_info}")
+
+            data = response.json()
+            for child in data.get("value", []):
+                new_path = path + "/" + child["name"]
+                if child.get("file"):  # Object is a file
+                    last_modified = datetime.strptime(child["lastModifiedDateTime"], "%Y-%m-%dT%H:%M:%SZ")
+                    yield (new_path, child["@microsoft.graph.downloadUrl"], last_modified)
+                else:  # Object is a folder, retrieve children
+                    child_url = f"{base_url}/items/{child['id']}/children"  # Use item endpoint for nested objects
+                    yield from get_files(child_url, new_path)
+            yield from []
```

Comment:
This logic is pretty much the same as in `list_directories_and_files`. Should we unify this logic to avoid repetition?

Reply:
Yes, the logic is the same, but this is just iterating over the tree. I don't think that we need to combine these functions.